File: scala/152.scala
object Solution {
    def maxProduct(nums: Array[Int]): Int = {
        (1 until nums.length).foldLeft((nums.head, nums.head, nums.head)) {
            case ((min, max, res), i) => {
                val n = nums(i)
                val newMin = n.min(min*n).min(max*n)
                val newMax = n.max(max*n).max(min*n)
                (newMin, newMax, res.max(newMax))
            }
        }._3
    }
}

File: leetcodeJava/152.java
package leetcodeJava;

class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0], min = 1, max = 1;
        for (var n : nums) {
            var newMin = Math.min(Math.min(min*n, n), max*n);
            var newMax = Math.max(Math.max(max*n, n), min*n);
            min = newMin;
            max = newMax;
            res = Math.max(res, newMax);
        }
        return res;
    }
}
